ip_reader.c

```diff
@@ -0,0 +1,257 @@
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+
+//consider padding
+struct ip4_address {
+	uint32_t addr;
+	uint32_t mask;
+	uint32_t network;
+	uint32_t broadcast;
+	char scope;
+	char broadc;
+	char loopb;
+	char reservd;
+	char class;
+};
+
+#define OCT1(ip) ((ip) >> 24)
+#define OCT2(ip) OCT1((ip) << 8)
+#define OCT3(ip) OCT1((ip) << 16)
+#define OCT4(ip) OCT1((ip) << 24)
+#define DOT_FORMAT "%u.%u.%u.%u"
+#define DOT_DECM(ip) OCT1(ip),OCT2(ip),OCT3(ip),OCT4(ip)
+
+/*
+ * checks validity of a given octet, used during translation of input when
+ * called by read_data()
+ * returns 0 on success and an error code on failure
+ */
+int value_valid(int);
+
+/*
+ * checks validity of a given address in text format, called by read_data()
+ * before translation
+ * returns 0 on success and an error code on failure
+ */
+int format_valid(char *);
+
+/*
+ * translates text input into a numeric address, can be used to translate
+ * an ipv4 address or mask, depending on given mode (0 - address, 1 - mask)
+ * returns 0 on success and an error code on failure
+ */
+int read_data(struct ip4_address *, char *, int);
+
+/*
+ * translates a prefix mask attached to an ip address input in text form into a
+ * numeric mask or infers a mask form address class.
+ * returns 0 on success and an error code on failure
+ */
+int demask(struct ip4_address *, char *);
+
+/*
+ * assigns a class to an address based on its first octet
+ */
+void get_network_class(struct ip4_address *);
+
+/*
+ * determines the scope of the address between private (1) and public (0)
+ */
+void get_scope(struct ip4_address *);
+
+/*
+ * prints to standard output relevant information that can be inferred form
+ * a given ip address and the optionally provided mask
+ */
+void get_info(struct ip4_address *);
+
+
+int main(int argc, char *argv[])
+{
+	struct ip4_address ip4 = {.addr = 0, .mask = 0};
+
+	if(read_data(&ip4, argv[1], 1)) {
+		printf("Invalid IP address!\n");
+		return 1;
+	}
+	get_network_class(&ip4);
+
+	switch (argc) {		//a lot of lines repeat, streamline
+	case 2:
+		if (demask(&ip4, argv[1])) {
+			printf("Not a valid subnet mask!\n");
+			return 1;
+		}
+		break;
+	case 3:
+		if (read_data(&ip4, argv[2], 0)) {
+			printf("Not a valid subnet mask!\n");
+			return 1;
+		}
+		break;
+	default:
+		printf("Invalid Input.\n");
+		return 1;
+	}
+
+	get_info(&ip4);
+	return 0;
+}
+
+int read_data(struct ip4_address *address, char *input, int mode)
+{
+	if (format_valid(input)) return 1;
+
+	long carry;	//contains currently read octed
+	char *end;	//tracks current octed read
+
+	for (int i = 0; i < 4; i++)
+	{
+		carry = strtol(input, &end, 10);
+		if (value_valid(carry)) {
```

Comment:
Not descriptive enough and in fact I'd prefer if the check was here. It's not that long.

Reply:
Done.